file/src/main/scala/akka/stream/alpakka/file/impl/archive/TarReaderStage.scala

```diff
@@ -37,70 +37,61 @@ private[file] class TarReaderStage
 
-      val ignoreFlowOutPull: OutHandler = new OutHandler {
-        override def onPull(): Unit = ()
-      }
-
-      val expectFlowPull: OutHandler = new OutHandler {
-        override def onPull(): Unit = {
-          pull(flowIn)
-          setHandler(flowOut, ignoreFlowOutPull)
-        }
-      }
-      val failOnFlowPush: InHandler = new InHandler {
-        override def onPush(): Unit = failStage(new TarReaderException("upstream pushed"))
-        override def onUpstreamFinish(): Unit = setKeepGoing(true)
-      }
-
-      setHandler(flowIn, readHeader(ByteString.empty))
-      setHandler(flowOut, expectFlowPull)
+      readHeader(ByteString.empty)
 
-      def readHeader(buffer: ByteString): InHandler = {
+      def readHeader(buffer: ByteString): Unit = {
         if (buffer.length >= TarArchiveEntry.headerLength) {
           readFile(buffer)
-        } else new CollectHeader(buffer)
+        } else setHandlers(flowIn, flowOut, new CollectHeader(buffer))
       }
 
-      def readFile(headerBuffer: ByteString): InHandler = {
-        def pushSource(metadata: TarArchiveMetadata, buffer: ByteString): InHandler = {
+      def readFile(headerBuffer: ByteString): Unit = {
+        def pushSource(metadata: TarArchiveMetadata, buffer: ByteString): Unit = {
           if (buffer.length >= metadata.size) {
             val (emit, remain) = buffer.splitAt(metadata.size.toInt)
             log.debug(s"emitting completed source for $metadata")
             push(flowOut, metadata -> Source.single(emit))
             readTrailer(metadata, remain, subSource = None)
-          } else new CollectFile(metadata, buffer)
+          } else setHandlers(flowIn, flowOut, new CollectFile(metadata, buffer))
         }
 
         if (headerBuffer.head == 0) {
           log.debug("empty filename, detected EOF padding, completing")
           complete(flowOut)
-          new FlushEndOfFilePadding()
+          setHandlers(flowIn, flowOut, new FlushEndOfFilePadding())
         } else {
           val metadata = TarArchiveEntry.parse(headerBuffer)
           val buffer = headerBuffer.drop(TarArchiveEntry.headerLength)
           if (isAvailable(flowOut)) {
             pushSource(metadata, buffer)
           } else {
-            // await flow demand
-            setHandler(flowOut, new OutHandler {
-              override def onPull(): Unit = {
-                setHandler(flowIn, pushSource(metadata, buffer))
+            setHandlers(
+              flowIn,
+              flowOut,
+              new OutHandler with InHandler {
+                override def onPull(): Unit = {
+                  setHandler(flowOut, IgnoreDownstreamPull)
+                  pushSource(metadata, buffer)
+                }
+
+                // fail on upstream push
+                override def onPush(): Unit = failStage(new TarReaderException("upstream pushed unexpectedly"))
+                override def onUpstreamFinish(): Unit = setKeepGoing(true)
               }
```

Comment:
I suggest giving this handler a name like we have for the others.

Reply:
👍